{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate enrichment scores across tree\n",
    "\n",
    "This is an initial attempt to break this code into reasonable chunks that can be coded, tested, and executed separately. In this notebook, we will be reading in a tree, enumerating all the mutations on that tree, and calculating enrichment scores for each of them. These enrichment scores are based on [this code](https://github.com/sheppardlab/pGWAS/blob/master/assomap_given_phylo.py), written for [this paper](https://www.nature.com/articles/s41467-018-07368-7#Sec10) detailed in lines 245-273 and calculated from the following contingency table as: \n",
    "\n",
    "|host|presence|absence|\n",
    "|:------|:-------|:------|\n",
    "|host 1|A|B| \n",
    "|host 2|C|D|\n",
    "\n",
    "where A, B, C, and D are counts of the mutation's presence and absence in host 1 and host 2. The odds ratio is then calculated as: `OR = (A * D)/(B * C)`\n",
    "\n",
    "In this notebook, this code is written for parsing a tree json format, output from Nextstrain. In subsequent notebooks, I will alter this for running on beast trees. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Infer each mutation that occurs across the tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"read in a tree in json format\"\"\"\n",
    "def read_in_tree_json(tree_path):\n",
    "    with open(tree_path) as json_file:\n",
    "        tree_json = json.load(json_file)\n",
    "\n",
    "    # Nextstrain tree jsons are broken into 2 components: metadata and tree. We just need the tree\n",
    "    tree_object=tree_json['tree']\n",
    "    meta=tree_json['meta']\n",
    "    json_translation={'absoluteTime':lambda k: k.traits['node_attrs']['num_date']['value'],'name':'name'} ## allows baltic to find correct attributes in JSON, height and name are required at a minimum\n",
    "\n",
    "    tree=bt.loadJSON(tree_object,json_translation)\n",
    "    \n",
    "    return(tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"count the number of tips on the tree corresponding to each host category\"\"\"\n",
    "def return_all_host_tips(tree, host1, host2, host_annotation):\n",
    "    host_counts = {host1:0, host2:0, \"other\":0}\n",
    "    \n",
    "    for k in tree.Objects: \n",
    "        if k.branchType == \"leaf\":\n",
    "            host = k.traits['node_attrs'][host_annotation]['value']\n",
    "            if host in [host1, host2]:\n",
    "                host_counts[host] += 1\n",
    "            else:\n",
    "                host_counts['other'] += 1\n",
    "    return(host_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"this function traverses the tree, from root to tip, and records 2 quantites:\n",
    "1. for each branch, it records the branch name and its branch length in a \n",
    "dictionary. When parsing a tree in baltic, each branch gets assigned a random, but unique name/identifier\n",
    "2. it adds up the total branch length on the tree as an integer\"\"\"\n",
    "\n",
    "def return_total_tree_branch_length(tree):\n",
    "    total_branch_length = 0\n",
    "    branch_lengths = {}\n",
    "    \n",
    "    for k in tree.Objects:\n",
    "        divergence = k.traits['node_attrs']['div']\n",
    "\n",
    "        # if this happens at the root, set parent divergence to 0\n",
    "        if k.parent.traits == {}:\n",
    "            parent_div = 0\n",
    "        else:\n",
    "            parent_div = k.parent.traits['node_attrs']['div']\n",
    "        \n",
    "        total_branch_length += divergence - parent_div\n",
    "        branch_lengths[k] = divergence-parent_div\n",
    "    \n",
    "    return(total_branch_length, branch_lengths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Traverse the tree from root to tip. On each branch, tips and nodes, gather every nucleotide and amino acid \n",
    "mutation\"\"\"\n",
    "\n",
    "def gather_all_mut_on_tree(tree, gene):\n",
    "    \n",
    "    all_aa_muts = []\n",
    "    all_nt_muts = []\n",
    "    \n",
    "    for k in tree.Objects:\n",
    "        if 'mutations' in k.traits['branch_attrs']:\n",
    "            if 'nuc' in k.traits['branch_attrs']['mutations']:\n",
    "                nt_muts = k.traits['branch_attrs']['mutations']['nuc']\n",
    "                all_nt_muts.extend(nt_muts)\n",
    "                \n",
    "            if gene in k.traits['branch_attrs']['mutations']:                \n",
    "                aa_muts = k.traits['branch_attrs']['mutations'][gene]\n",
    "                all_aa_muts.extend(aa_muts)\n",
    "    \n",
    "    # subset to only unique mutations \n",
    "    all_aa_muts = list(set(all_aa_muts))\n",
    "    all_nt_muts = list(set(all_nt_muts))\n",
    "    \n",
    "    return(all_aa_muts, all_nt_muts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"return the total number of times that the mutation arises on the phylogeny. This includes instances of mutation \n",
    "on internal nodes and on tips and counts each with the same weight\"\"\"\n",
    "\n",
    "def return_number_times_on_tree(tree, mut, gene):\n",
    "    times_on_tree = 0\n",
    "    \n",
    "    for k in tree.Objects:\n",
    "        if 'mutations' in k.traits['branch_attrs']:\n",
    "        \n",
    "            if gene in k.traits['branch_attrs']['mutations']:\n",
    "                aa_muts = k.traits['branch_attrs']['mutations'][gene]\n",
    "                if mut in aa_muts: \n",
    "                    times_on_tree += 1\n",
    "                                                \n",
    "    return(times_on_tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"return the total branch length for all branches on which the mutation arises, added together. This includes \n",
    "instances of mutation on internal nodes and on tips\"\"\"\n",
    "\n",
    "def return_branch_length_mut_on_tree(tree, mut, gene):\n",
    "    branch_length = 0\n",
    "            \n",
    "    for k in tree.Objects:\n",
    "        if 'mutations' in k.traits['branch_attrs']:\n",
    "            if gene in k.traits['branch_attrs']['mutations']:\n",
    "                aa_muts = k.traits['branch_attrs']['mutations'][gene]\n",
    "                \n",
    "                # if the mutation arises on this branch\n",
    "                if mut in aa_muts: \n",
    "                    divergence = k.traits['node_attrs']['div']\n",
    "                    \n",
    "                    # if this happens at the root, set parent divergence to 0\n",
    "                    if k.parent.traits == {}:\n",
    "                        parent_div = 0\n",
    "                    elif 'node_attrs' not in k.parent.traits:\n",
    "                        parent_div = 0\n",
    "                    else:\n",
    "                        parent_div = k.parent.traits['node_attrs']['div']\n",
    "                \n",
    "                    local_branch_length = divergence - parent_div\n",
    "                    branch_length += local_branch_length\n",
    "                    \n",
    "    return(branch_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"given a branch and gene, return the mutations present on that branch\"\"\"\n",
    "\n",
    "def return_muts_on_branch(branch, gene):\n",
    "    muts = []\n",
    "    if 'branch_attrs' in branch.traits:\n",
    "        if 'mutations' in branch.traits['branch_attrs']:\n",
    "            if gene in branch.traits['branch_attrs']['mutations']:\n",
    "                muts = branch.traits['branch_attrs']['mutations'][gene]\n",
    "                            \n",
    "    return(muts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Given a starting internal node, and a tip you would like to end at, traverse the full path from that node to\n",
    "tip. Along the way, gather mutations that occur along that path. Once you have reached the ending \n",
    "tip, return the list of mutations that fell along that path\"\"\"\n",
    "\n",
    "def return_all_muts_on_path_to_tip(starting_node, ending_tip, gene, muts, host_annotation):\n",
    "    \n",
    "    # set an empty list of mutations and enumerate the children of the starting node; children can be tips or nodes\n",
    "    children = starting_node.children\n",
    "    \n",
    "    for child in children:\n",
    "        local_muts = []\n",
    "        \n",
    "        \"\"\"if the child is a leaf: if leaf is the target end tip, add the mutations that occur on that branch to \n",
    "        the list and return the list; if leaf is not the target end tip, move on\"\"\"\n",
    "        \"\"\"if the child is an internal node: first, test whether that child node contains the target tips in its \n",
    "        children. child.leaves will output a list of the names of all tips descending from that node. If not, pass. \n",
    "        if the node does contain the target end tip in its leaves, keep traversing down that node recursively, \n",
    "        collecting mutations as you go\"\"\"\n",
    "\n",
    "        if child.branchType == \"leaf\":\n",
    "            if child.name != ending_tip:\n",
    "                pass\n",
    "            elif child.name == ending_tip:\n",
    "                host = child.traits['node_attrs'][host_annotation]['value']\n",
    "                local_muts = return_muts_on_branch(child, gene)\n",
    "                muts.extend(local_muts)\n",
    "                #print(child.name, host, muts)\n",
    "                return(host, muts)        \n",
    "        \n",
    "        elif child.branchType == \"node\":\n",
    "            if ending_tip not in child.leaves:\n",
    "                pass\n",
    "            else:\n",
    "                local_muts = return_muts_on_branch(child, gene)\n",
    "                muts.extend(local_muts)\n",
    "                host, muts = return_all_muts_on_path_to_tip(child, ending_tip, gene, muts,host_annotation)\n",
    "    \n",
    "    return(host, muts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"at times, will need to check whether the revertant mutation occcurs downstream. Return the revertant mutation\"\"\"\n",
    "\n",
    "def return_opposite_mutation(mut):\n",
    "    \n",
    "    site = mut[1:-1]\n",
    "    ref = mut[0]\n",
    "    alt = mut[-1]\n",
    "    opposite = alt+site+ref\n",
    "    \n",
    "    return(opposite)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"given a tree, mutation, and gene, return the number of times that mutation is present in each host\"\"\"\n",
    "\n",
    "def return_host_distribution_mutation(tree, mut, gene, host1, host2, host_annotation):\n",
    "    \n",
    "    host_counts_dict = {host1:0, host2:0, \"other\":0}\n",
    "    back_mutation = return_opposite_mutation(mut)\n",
    "    \n",
    "    # iterate through tree\n",
    "    for k in tree.Objects:\n",
    "        if 'mutations' in k.traits['branch_attrs']:\n",
    "            if gene in k.traits['branch_attrs']['mutations']:\n",
    "                aa_muts = k.traits['branch_attrs']['mutations'][gene]  \n",
    "                \n",
    "                # if we have reached a node or tip in the tree with the target mutation, enumerate descendants\n",
    "                if mut in aa_muts: \n",
    "                    \n",
    "                    # if the mutation occurs on a leaf, record the host and move on \n",
    "                    if k.branchType == 'leaf':\n",
    "                        host = k.traits['node_attrs'][host_annotation]['value']\n",
    "                        if host in [host1,host2]:\n",
    "                            host_counts_dict[host] += 1\n",
    "                        else:\n",
    "                            host_counts_dict[\"other\"]+= 1\n",
    "                    \n",
    "                    # else, if the mutation occurs on a node, traverse the children and return host\n",
    "                    elif k.branchType == \"node\":\n",
    "                        all_leaves = k.leaves\n",
    "                        for leaf in all_leaves: \n",
    "                            muts = []\n",
    "                            host, muts = return_all_muts_on_path_to_tip(k, leaf, gene, muts, host_annotation)\n",
    "                            if back_mutation in muts: \n",
    "                                pass\n",
    "                            elif back_mutation and mut in muts:  # if both the mutation and backmutation occur, print\n",
    "                                print(leaf, back_mutation, mut)\n",
    "                            else:\n",
    "                                if host in [host1, host2]:\n",
    "                                    host_counts_dict[host] += 1\n",
    "                                else: \n",
    "                                    host_counts_dict[\"other\"] += 1\n",
    "                                \n",
    "    return(host_counts_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculate the enrichment scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"calculate an enrichment score for an individual mutation, based on the counts across hosts\"\"\"\n",
    "\n",
    "def calculate_enrichment_score_counts(mut_counts_dict, host1, host2, host_counts):\n",
    "    total_host1_tree = host_counts[host1]\n",
    "    total_host2_tree = host_counts[host2]\n",
    "    \n",
    "    mut_host1 = mut_counts_dict[host1]\n",
    "    mut_host2 = mut_counts_dict[host2]\n",
    "    \n",
    "    # this is calculating this table as counts\n",
    "    presence_host1 = mut_host1\n",
    "    absence_host1 = total_host1_tree - mut_host1\n",
    "    presence_host2 = mut_host2\n",
    "    absence_host2 = total_host2_tree - mut_host2\n",
    "    \n",
    "    if presence_host2 == 0:\n",
    "        presence_host2 = 1\n",
    "    if absence_host1 == 0:\n",
    "        absence_host1 = 1\n",
    "\n",
    "    # this score is calculated in terms of its enrichment in host 1\n",
    "    score = (presence_host1 * absence_host2)/(presence_host2 * absence_host1)\n",
    "    from scipy.stats import fisher_exact\n",
    "    oddsr, p = fisher_exact([[presence_host1, absence_host1],[presence_host2, absence_host2]], alternative='two-sided')\n",
    "    \n",
    "    return(oddsr, p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"calculate an enrichment score for an individual mutation, based on the counts across hosts\"\"\"\n",
    "\n",
    "def calculate_enrichment_score_proportions(mut_counts_dict, host1, host2, host_counts):\n",
    "    total_host1_tree = host_counts[host1]\n",
    "    total_host2_tree = host_counts[host2]\n",
    "    \n",
    "    mut_host1 = mut_counts_dict[host1]\n",
    "    mut_host2 = mut_counts_dict[host2]\n",
    "    \n",
    "    total_tips_in_tree = total_host1_tree + total_host2_tree\n",
    "    \n",
    "    # this is calculating this table as proportions\n",
    "    presence_host1 = (mut_host1)/total_tips_in_tree\n",
    "    absence_host1 = (total_host1_tree - mut_host1)/total_tips_in_tree\n",
    "    presence_host2 = (mut_host2)/total_tips_in_tree\n",
    "    absence_host2 = (total_host2_tree - mut_host2)/total_tips_in_tree\n",
    "    \n",
    "#     if presence_host2 == 0:\n",
    "#         presence_host2 = 1\n",
    "#     if absence_host1 == 0:\n",
    "#         absence_host1 = 1\n",
    "\n",
    "    # this score is calculated in terms of its enrichment in host 1\n",
    "    #score = (presence_host1 * absence_host2)/(presence_host2 * absence_host1)\n",
    "    score = (presence_host1 + absence_host2) - (presence_host2 + absence_host1)\n",
    "    return(score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"for a tree and all amino acid mutations, calculate the enrichment scores across the tree\"\"\"\n",
    "def calculate_enrichment_scores(tree, aa_muts, nt_muts, host1, host2, host_annotation, min_required_count, method, host_counts):\n",
    "    scores_dict = {}\n",
    "    times_detected_dict = {}\n",
    "    branch_lengths_dict = {}\n",
    "    host_counts_dict2 = {}\n",
    "    \n",
    "    if method == \"counts\":\n",
    "        enrichment_calculation_function = calculate_enrichment_score_counts\n",
    "    elif method == \"proportions\":\n",
    "        enrichment_calculation_function = calculate_enrichment_score_proportions\n",
    "    \n",
    "    for a in aa_muts:\n",
    "        times_detected = return_number_times_on_tree(tree, a, gene)\n",
    "        times_detected_dict[a] = times_detected\n",
    "\n",
    "        branch_length_mut = return_branch_length_mut_on_tree(tree, a, gene)\n",
    "        branch_lengths_dict[a] = branch_length_mut\n",
    "\n",
    "        host_counts_dict = return_host_distribution_mutation(tree, a, gene, host1, host2, host_annotation)\n",
    "        host_counts_dict2[a] = host_counts_dict\n",
    "        total_tips_with_mut = host_counts_dict[host1] + host_counts_dict[host2] + host_counts_dict[\"other\"]\n",
    "\n",
    "        if total_tips_with_mut >= min_required_count:\n",
    "            enrichment_score,p_value = enrichment_calculation_function(host_counts_dict, host1, host2, host_counts)\n",
    "            #print(enrichment_score, total_tips_with_mut, host_counts_dict)\n",
    "            scores_dict[a] = {\"enrichment_score\": enrichment_score, \"pvalue\":p_value}\n",
    "            \n",
    "            \n",
    "    return(scores_dict, times_detected_dict, branch_lengths_dict, host_counts_dict2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "LHM-basics (python3)",
   "language": "python",
   "name": "lhm-basics"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
