{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analyze mutations on host switch branches\n",
    "\n",
    "As another test for host-switching associated mutations, I would like to do this branch test to determine whether there are mutations that are associated specifically with cross-species transmission events. In this notebook, I will do the following: \n",
    "\n",
    "1. For each branch on the tree, determine whether it is part of a host switch or not. A branch is part of a host switch if it's host state is different from its parent host state, or if it's child branch is a different host state than it is. \n",
    "2. If the branch is part of a host switch, categorize whether it is the parent or child branch. In this mascot model with the migration events mapped onto individual branches, what you end up with is for host switch branches, 2 branches of distinct host states that are continuous. So for example, the parent branch will have one child, which is the portion of the entire branch which is a different host state. \n",
    "3. Enumerate whether there are any mutations on the branch. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import re,copy, imp\n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "# for this to work, you will need to download the most recent version of baltic, available here \n",
    "bt = imp.load_source('baltic', '../baltic/baltic/baltic-modified-for-muts.py')\n",
    "from io import StringIO\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "from datetime import date\n",
    "current_date = str(date.today())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define colors \n",
    "domestic_color=\"#4E83AE\"\n",
    "wild_color=\"#CEB540\"\n",
    "human_color=\"#DE4428\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Translations and ancestral sequence reconstruction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_alignment(alignment_file):\n",
    "    alignment_dict = {}\n",
    "\n",
    "    for seq in SeqIO.parse(alignment_file, \"fasta\"):\n",
    "        seqName = seq.description \n",
    "        sequence = str(seq.seq)\n",
    "        alignment_dict[seqName] = sequence\n",
    "        \n",
    "    return(alignment_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "def return_cds_coordinates(genbank_ref_file):\n",
    "    \n",
    "    from Bio import GenBank\n",
    "    with open(genbank_ref_file) as handle:\n",
    "        for record in GenBank.parse(handle):\n",
    "\n",
    "            # pull out the CDS feature; the gene coordinatees are in the feature.location. Get help with help(feauture)\n",
    "            for f in record.features:\n",
    "                if f.key == \"CDS\":\n",
    "                    cds_start = int(f.location.split(\"..\")[0])\n",
    "                    cds_stop = int(f.location.split(\"..\")[1])\n",
    "                \n",
    "    return(cds_start, cds_stop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "def return_mutations_on_branch(branch):\n",
    "    if branch == None: \n",
    "        mutations = []\n",
    "    elif \"mutations\" in branch.traits: \n",
    "        mutations = branch.traits[\"mutations\"].split(\",\")\n",
    "    else:\n",
    "        mutations = []\n",
    "    \n",
    "    return(mutations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "def return_mutated_sequence(sequence, muts, cds_start, cds_stop):\n",
    "    # make into a list because strings are immutable, while list are not\n",
    "    mutated_sequence = list(sequence)\n",
    "    \n",
    "    for m in muts:\n",
    "        site = int(m[1:-1])-1   # -1 is because of 0 indexing\n",
    "        ancestral_nt = m[0]\n",
    "        mutated_nt = m[-1]\n",
    "        \n",
    "        # since we are going backwards up the tree, we are reconstructing the ancestral sequence\n",
    "        mutated_sequence[site] = ancestral_nt\n",
    "    \n",
    "    mutated_sequence = \"\".join(mutated_sequence)\n",
    "    mutated_aa_sequence, aa_muts = return_mutated_aa_sequence(sequence, mutated_sequence, cds_start, cds_stop)\n",
    "    # return a string\n",
    "    return(mutated_sequence, mutated_aa_sequence, aa_muts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "def return_aa_sequence(sequence, cds_start, cds_stop):\n",
    "    \n",
    "    ha_cds = str(sequence)[cds_start-1:cds_stop-1]    # slice string based on cds coordinates\n",
    "    ha_cds_seq = Seq(ha_cds)    # make it a Seq object\n",
    "    translation = ha_cds_seq.translate()\n",
    "    \n",
    "    return(str(translation))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "def return_mutated_aa_sequence(sequence, mutated_sequence, cds_start, cds_stop):\n",
    "    \n",
    "    ha_cds = str(sequence)[cds_start-1:cds_stop-1]    # slice string based on cds coordinates\n",
    "    ha_cds_seq = Seq(ha_cds)    # make it a Seq object\n",
    "    translation = ha_cds_seq.translate()\n",
    "    mutated_translation = str(Seq(str(mutated_sequence)[cds_start-1:cds_stop-1]).translate())  # same as above but on 1 line\n",
    "    \n",
    "    aa_muts = []\n",
    "    for i in range(len(translation)):\n",
    "        if mutated_translation[i] != translation[i]:\n",
    "            aa_mut = mutated_translation[i] + str(i+1) + translation[i]\n",
    "            aa_muts.append(aa_mut)\n",
    "\n",
    "    return(mutated_translation, aa_muts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "def return_all_parents(k, parents_dict, sequence, cds_start, cds_stop):\n",
    "    mutations = return_mutations_on_branch(k)\n",
    "    \n",
    "    # if at root\n",
    "    if k.parent == None:\n",
    "        return(parents_dict)\n",
    "    \n",
    "    # if not at root yet\n",
    "    elif k.branchType == \"leaf\":\n",
    "        \n",
    "        # do something else here....we've already recorded the mutations and stuff so we should just go up one\n",
    "        parents_dict = return_all_parents(k.parent, parents_dict, sequence, cds_start, cds_stop)\n",
    "    \n",
    "    else:\n",
    "        sequence, aa_sequence, aa_muts = return_mutated_sequence(sequence, mutations, cds_start, cds_stop)\n",
    "        parents_dict[k] = {\"nt_muts\": mutations, \"sequence\":sequence, \"aa_sequence\":aa_sequence, \n",
    "                                  \"aa_muts\":aa_muts}\n",
    "        parents_dict = return_all_parents(k.parent, parents_dict, sequence, cds_start, cds_stop)\n",
    "        \n",
    "    return(parents_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "def return_sequence_map(tree, alignment_dict, cds_start, cds_stop):\n",
    "    \n",
    "    all_nodes = {}\n",
    "\n",
    "    for k in tree.Objects: \n",
    "        if k.branchType == \"leaf\":\n",
    "            sequence = alignment_dict[k.name]\n",
    "            aa_sequence = return_aa_sequence(sequence, cds_start, cds_stop)\n",
    "            mutations = return_mutations_on_branch(k)\n",
    "            mutated_sequence, mutated_aa_sequence, aa_muts = return_mutated_sequence(sequence, mutations, cds_start, cds_stop)\n",
    "            all_nodes[k] = {\"muts\":mutations, \"aa_muts\":aa_muts, \"leaves\":\"NA\"}\n",
    "            \n",
    "            # parents dict will include all parental nodes from the tip back to the root with their mutations, \n",
    "            # nucleotide sequences, and names as 'branchName':{'nt_muts':[list of nt muts], 'sequence': str(nt sequence)}\n",
    "            parents_dict = {}\n",
    "            parents_dict = return_all_parents(k, parents_dict, sequence, cds_start, cds_stop)\n",
    "            \n",
    "            # make a master list of internal nodes we've already inferred to not repeat work\n",
    "            for p in parents_dict:\n",
    "                leaves = p.leaves\n",
    "                sequence = parents_dict[p]['sequence']\n",
    "                aa_sequence = parents_dict[p]['aa_sequence']\n",
    "                aa_muts = parents_dict[p]['aa_muts']\n",
    "                muts = parents_dict[p]['nt_muts']\n",
    "                \n",
    "                # check to see if the name matches and if the leaves match; sometimes baltic assigns the same numeric\n",
    "                # name to 2 different nodes!\n",
    "                if p in all_nodes:\n",
    "                    if all_nodes[p]['leaves'] == leaves:\n",
    "                        pass\n",
    "                    else:\n",
    "                        all_nodes[p] = {\"muts\":muts, \"aa_muts\":aa_muts, \"leaves\":leaves}\n",
    "                else: \n",
    "                    all_nodes[p] = {\"muts\":muts, \"aa_muts\":aa_muts, \"leaves\":leaves}\n",
    "\n",
    "    return(all_nodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Grab all trees and process posterior"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_taxa_lines(tree_path):    \n",
    "\n",
    "    lines_to_write = \"\"\n",
    "    with open(tree_path, 'rU') as infile:\n",
    "        for line in infile: ## iterate through each line\n",
    "            if 'state' not in line.lower(): #going to grab all the interesting stuff in the .trees file prior to the newick tree strings\n",
    "                lines_to_write = lines_to_write + line\n",
    "\n",
    "    return(lines_to_write)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_strain_to_number(taxa_lines):\n",
    "    \n",
    "    output_dict = {}\n",
    "    \n",
    "    translation_block = taxa_lines.split(\"Translate\\n\")[1]\n",
    "    translation_list = translation_block.replace(\"\\t\",\"\").split(\"\\n\")\n",
    "    \n",
    "    for t in translation_list: \n",
    "        information = t.lstrip().replace(\",\",\"\")  # remove leading white spaces and commas\n",
    "        \n",
    "        if len(information.split(\" \")) == 2:\n",
    "            numeric_id = information.split(\" \")[0]\n",
    "            strain_name = information.split(\" \")[1]\n",
    "        \n",
    "            output_dict[numeric_id] = strain_name\n",
    "            \n",
    "        else: \n",
    "            pass\n",
    "        \n",
    "    return(output_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_leaves_to_strains(input_leaves, strains_dict): \n",
    "    output_list = []\n",
    "    \n",
    "    for l in input_leaves: \n",
    "        strain_name = strains_dict[l]\n",
    "        output_list.append(strain_name)\n",
    "    \n",
    "    output_set = set(output_list)\n",
    "    return(output_set)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_burnin_value(tree_path, burnin_percent):\n",
    "    with open(tree_path, 'rU') as infile:\n",
    "        numtrees = 0\n",
    "        for line in infile: ## iterate through each line\n",
    "            if 'state' in line.lower(): #going to grab all the interesting stuff in the .trees file prior to the newick tree strings\n",
    "                numtrees += 1\n",
    "    \n",
    "    burnin = numtrees * burnin_percent\n",
    "    return(burnin)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"this is a function I'm not currently using, but it will enumerate mutations as well as which host type branches\n",
    "they occur on\"\"\"\n",
    "\n",
    "def enumerate_mutations(tree):\n",
    "    muts = {}\n",
    "    host_branches = {}\n",
    "    \n",
    "    for k in tree.Objects:\n",
    "        host = k.traits['typeTrait']\n",
    "        if host in host_branches: \n",
    "            host_branches[host] += 1\n",
    "        else:\n",
    "            host_branches[host] = 1\n",
    "        \n",
    "        if 'mutations' in k.traits: \n",
    "            mutations = k.traits['mutations'].split(\",\")\n",
    "            \n",
    "            for m in mutations: \n",
    "                site = m[1:-1]\n",
    "                if site not in muts: \n",
    "                    muts[site] = [host]\n",
    "                else: \n",
    "                    muts[site].append(host)\n",
    "            \n",
    "    return(muts, host_branches)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tabulate_mutations(mutations_dict):\n",
    "    output_dict = {}\n",
    "    for mut in mutations_dict:\n",
    "        output_dict[mut] = {}\n",
    "        hosts = list(set(mutations_dict[mut]))\n",
    "        for h in hosts: \n",
    "            output_dict[mut][h] = mutations_dict[mut].count(h)\n",
    "    \n",
    "    return(output_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "def muts_on_branch(branch):\n",
    "    \n",
    "    if 'mutations' in branch.traits: \n",
    "        mutations = branch.traits['mutations'].split(\",\")\n",
    "    \n",
    "    else: \n",
    "        mutations = \"\"\n",
    "    \n",
    "    return(mutations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"given a branch, determine whether any of its children are of a different host state\"\"\"\n",
    "\n",
    "def enumerate_child_traits(branch):\n",
    "    child_traits = []\n",
    "    \n",
    "    current_trait = branch.traits['typeTrait']\n",
    "    \n",
    "    for c in branch.children: \n",
    "        child_trait = c.traits['typeTrait']\n",
    "        child_traits.append(child_trait)\n",
    "    \n",
    "    if set(child_traits) == {current_trait}:\n",
    "        host_switch_branch = \"no\"\n",
    "    \n",
    "    else:\n",
    "        host_switch_branch = \"yes\"\n",
    "        \n",
    "    return(host_switch_branch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"given a parent node and its child, infer whether either the parent or child has a mutation on it\"\"\"\n",
    "\n",
    "def mut_near_migration_event(parent, child):\n",
    "    \n",
    "    if 'mutations' in parent.traits:\n",
    "        parent_mutations = parent.traits['mutations'].split(\",\")\n",
    "    else:\n",
    "        parent_mutations = []\n",
    "        \n",
    "    if 'mutations' in child.traits:\n",
    "        child_mutations = child.traits['mutations'].split(\",\")\n",
    "        \n",
    "    else:\n",
    "        child_mutations = []\n",
    "        \n",
    "    return(parent_mutations, child_mutations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"this function will iterate through the tree, and record each inferred migration event along the phylogeny. It\n",
    "will also capture the inferred time of the migration event as a decimal date, and whether the parent or child branch\n",
    "surrounding the migration event have any mutations on them. These values are stored in a dictionary where the \n",
    "key is a number to keep track of the total number of migration events on the tree, and the value is a dictionary\n",
    "housing data about the migration event. this includes its type, the parent and child host states, date, and mutations\"\"\"\n",
    "\n",
    "def enumerate_mutations_and_migration_events(tree, sequence_map):\n",
    "    output_dict = {}\n",
    "    branch_counter = 0\n",
    "        \n",
    "    for k in tree.Objects:\n",
    "                \n",
    "        trait = k.traits['typeTrait']\n",
    "        parent_node = k.parent\n",
    "        \n",
    "        if 'typeTrait' not in parent_node.traits:\n",
    "            parent_trait = \"root\"\n",
    "        \n",
    "        # only write out migration events that are not from root to deme\n",
    "        else:\n",
    "            parent_trait = parent_node.traits['typeTrait']\n",
    "                        \n",
    "            # if the current trait differs from the parent trait, then this is a host switch event. Enumerate \n",
    "            # mutations on the parent and current branch; here, I'm denoting the current branch as the child \n",
    "            # branch because I am labelling host switches as being comprised of a parent and child state/branch\n",
    "            if trait != parent_trait:\n",
    "                host_switch_event = \"yes\"\n",
    "                migration_event = parent_trait + \"-to-\" + trait\n",
    "\n",
    "                # does a mutation occur on the parent or child branch? \n",
    "                parent_mut, child_mut = mut_near_migration_event(parent_node, k)\n",
    "                \n",
    "                # write out the information for the parent and child branches\n",
    "                branch_counter += 1\n",
    "                if parent_node in sequence_map: \n",
    "                    parent_aa_muts = sequence_map[parent_node]['aa_muts']\n",
    "                else:\n",
    "                    print(str(parent_node) + \"not in sequence map\")\n",
    "                    \n",
    "                output_dict[branch_counter] = {\"type\":migration_event, \"host_switch\": host_switch_event,\n",
    "                                                 \"host\":parent_trait, \"role\":\"parent\", \"mutations\":parent_mut, \n",
    "                                               \"aa_muts\":parent_aa_muts,\"branch_id\":parent_node}\n",
    "\n",
    "                if k in sequence_map: \n",
    "                    aa_muts = sequence_map[k]['aa_muts']\n",
    "                else:\n",
    "                    print(str(k) + \"not in sequence map\")\n",
    "                branch_counter += 1\n",
    "                output_dict[branch_counter] = {\"type\":migration_event, \"host_switch\": host_switch_event,\n",
    "                                                 \"host\":trait, \"role\":\"child\", \"mutations\":child_mut, \n",
    "                                               \"aa_muts\":aa_muts,\"branch_id\":k}\n",
    "\n",
    "\n",
    "            \n",
    "            # The below if else business is to make sure that we don't count mutations on some branches twice\n",
    "            # If we have a non-host switch branch that will be the parent to a host switch, we have to make sure that \n",
    "            # that branch's mutations aren't counted twice \n",
    "            # if the current trait is the same as the parent trait:\n",
    "            else:\n",
    "                \n",
    "                # if the branch is a node and it has no children or if it is a leaf, then you are not a host \n",
    "                # switch branch, and we can record your mutations and be done\n",
    "                # here, we record the role as \"on\" because the mutations are just on the branch and we don't have\n",
    "                # parent and child branches\n",
    "                if (k.branchType == 'node' and len(k.children) == 0) or k.branchType == 'leaf':\n",
    "                    mutations = muts_on_branch(k)\n",
    "                    aa_muts = sequence_map[k][\"aa_muts\"]\n",
    "                    migration_event = \"within-\" + trait\n",
    "                    host_switch_event = \"no\"\n",
    "                    \n",
    "                    branch_counter += 1\n",
    "                    output_dict[branch_counter] = {\"type\":migration_event, \"host_switch\": host_switch_event,\n",
    "                                                 \"host\":trait, \"role\":\"on\", \"mutations\":mutations,\n",
    "                                                   \"aa_muts\":aa_muts,\"branch_id\":k}\n",
    "                \n",
    "                # if you are a node and you do have children, we need to check your children to make sure that \n",
    "                # they do not have a trait change. If they do, then you are part of a host switch.\n",
    "                elif len(k.children) > 0:\n",
    "                    host_switch_event = enumerate_child_traits(k)\n",
    "                    \n",
    "                    # if we are on a host_switch branch, pass; we will detect this when we get down to the child\n",
    "                    # branch that is a host switch \n",
    "                    if host_switch_event == \"yes\":\n",
    "                        to_record = \"no\"\n",
    "                    else: \n",
    "                        mutations = muts_on_branch(k)\n",
    "                        aa_muts = sequence_map[k]['aa_muts']\n",
    "                        migration_event = \"within-\" + trait\n",
    "                        \n",
    "                        branch_counter += 1\n",
    "                        output_dict[branch_counter] = {\"type\":migration_event, \"host_switch\": host_switch_event,\n",
    "                                                 \"host\":trait, \"role\":\"on\", \"mutations\":mutations, \n",
    "                                                       \"aa_muts\":aa_muts,\"branch_id\":k}\n",
    "                \n",
    "    return(output_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_on_posterior_trees(all_trees, burnin, genbank_ref_file, alignment):\n",
    "    start_time = time.time()\n",
    "\n",
    "    cds_start, cds_stop = return_cds_coordinates(genbank_ref_file)\n",
    "    alignment_dict = read_alignment(alignment)\n",
    "    \n",
    "    with open(all_trees, \"r\") as infile:\n",
    "\n",
    "        tree_counter = 0\n",
    "        mutations_dict = {}\n",
    "\n",
    "        for line in infile:\n",
    "            if 'tree STATE_' in line:\n",
    "                tree_counter += 1\n",
    "                if tree_counter >= burnin:\n",
    "\n",
    "                    temp_tree = StringIO(taxa_lines + line)\n",
    "                    tree = bt.loadNexus(temp_tree)\n",
    "                    #tree.setAbsoluteTime(2019.227)\n",
    "                    \n",
    "                    # generate the sequence map, which maps for each branch the mutations, aa muts, and sequences\n",
    "                    sequence_map = return_sequence_map(tree, alignment_dict, cds_start, cds_stop)\n",
    "\n",
    "                    # iterate through the tree and pull out all migration events\n",
    "                    mutations_dict[tree_counter] = enumerate_mutations_and_migration_events(tree, sequence_map)\n",
    "\n",
    "    # print the amount of time this took\n",
    "    total_time_seconds = time.time() - start_time\n",
    "    total_time_minutes = total_time_seconds/60\n",
    "    print(\"this took\", total_time_seconds, \"seconds (\", total_time_minutes,\" minutes) to run on\", tree_counter, \"trees\")\n",
    "    return(mutations_dict, sequence_map)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run on posterior trees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "# trees file paths\n",
    "skyline_human_off = {\"label\": \"skyline-human-off\", \"mig_direction\":\"forwards in time\",\n",
    "                      \"trees\": \"../../h5n1-host-classification/beast/beast-runs/2022-04-19-mascot-3deme-skyline-fixed-muts-logger/combined-2022-06-01.muts.trees\"}\n",
    "# skyline_human_off = {\"label\": \"skyline-human-off\", \"mig_direction\":\"forwards in time\",\n",
    "#                      \"trees\": \"/Volumes/data-backups-post-doc/stored_files_too_big_for_laptop/h5n1-host-classification/beast/beast-runs/2021-03-15-mascot-3deme-skyline-with-mig-history/no-human-mig-2021-03-31/2021-03-15-mascot-3deme-no-human-mig.combined.trees\"}\n",
    "\n",
    "alignment = \"../../h5n1-host-classification/beast/alignments/aligned_h5n1_ha-3deme-1per-country-month-host-downsampled-bad-dates-2021-06-09-with-annotations-2021-07-06.fasta\"\n",
    "genbank_ref_file = \"../test-data/reference_h5n1_ha.gb\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "to_run = skyline_human_off\n",
    "\n",
    "trees_file_path = to_run['trees']\n",
    "label = to_run['label']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/lmoncla/anaconda/envs/LHM-basics/lib/python3.7/site-packages/ipykernel_launcher.py:4: DeprecationWarning: 'U' mode is deprecated\n",
      "  after removing the cwd from sys.path.\n",
      "/Users/lmoncla/anaconda/envs/LHM-basics/lib/python3.7/site-packages/ipykernel_launcher.py:2: DeprecationWarning: 'U' mode is deprecated\n",
      "  \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "all_trees = trees_file_path\n",
    "burnin_percent = 0\n",
    "\n",
    "taxa_lines = get_taxa_lines(all_trees)\n",
    "burnin = get_burnin_value(all_trees, burnin_percent)\n",
    "print(burnin)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "this took 14966.451487064362 seconds ( 249.44085811773937  minutes) to run on 1133 trees\n"
     ]
    }
   ],
   "source": [
    "mutations_dict, sequence_map = run_on_posterior_trees(all_trees, burnin, genbank_ref_file, alignment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "#mutations_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tree_number</th>\n",
       "      <th>branch_number</th>\n",
       "      <th>type</th>\n",
       "      <th>host_switch</th>\n",
       "      <th>host</th>\n",
       "      <th>role</th>\n",
       "      <th>mutations</th>\n",
       "      <th>aa_muts</th>\n",
       "      <th>branch_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>within-wild</td>\n",
       "      <td>no</td>\n",
       "      <td>wild</td>\n",
       "      <td>on</td>\n",
       "      <td>[C72T, G173A, T392C, A856G, C1171A, C1341T]</td>\n",
       "      <td>[R51K, I124T, T279A, Q384K]</td>\n",
       "      <td>&lt;baltic.node object at 0x7f86187a5278&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>within-wild</td>\n",
       "      <td>no</td>\n",
       "      <td>wild</td>\n",
       "      <td>on</td>\n",
       "      <td>[A9C, C957T]</td>\n",
       "      <td>[]</td>\n",
       "      <td>&lt;baltic.node object at 0x7f86187a52b0&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>within-wild</td>\n",
       "      <td>no</td>\n",
       "      <td>wild</td>\n",
       "      <td>on</td>\n",
       "      <td>[T10C, A202G, C219A, C392T, T447A, G488A, C660...</td>\n",
       "      <td>[N61D, T124I, D142E, R156K, E228K, P233S]</td>\n",
       "      <td>&lt;baltic.node object at 0x7f86187a52e8&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>within-wild</td>\n",
       "      <td>no</td>\n",
       "      <td>wild</td>\n",
       "      <td>on</td>\n",
       "      <td>[T264A, T483G, C699A, C783T, G1287A, T1464C]</td>\n",
       "      <td>[H154Q]</td>\n",
       "      <td>&lt;baltic.node object at 0x7f86187a5320&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>within-wild</td>\n",
       "      <td>no</td>\n",
       "      <td>wild</td>\n",
       "      <td>on</td>\n",
       "      <td>[A672G, T789C, A1017G, C1028G, A1293G, A1517G,...</td>\n",
       "      <td>[T336S, K499R, M527I]</td>\n",
       "      <td>&lt;baltic.node object at 0x7f86187a5358&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   tree_number  branch_number         type host_switch  host role  \\\n",
       "0            1              1  within-wild          no  wild   on   \n",
       "1            1              2  within-wild          no  wild   on   \n",
       "2            1              3  within-wild          no  wild   on   \n",
       "3            1              4  within-wild          no  wild   on   \n",
       "4            1              5  within-wild          no  wild   on   \n",
       "\n",
       "                                           mutations  \\\n",
       "0        [C72T, G173A, T392C, A856G, C1171A, C1341T]   \n",
       "1                                       [A9C, C957T]   \n",
       "2  [T10C, A202G, C219A, C392T, T447A, G488A, C660...   \n",
       "3       [T264A, T483G, C699A, C783T, G1287A, T1464C]   \n",
       "4  [A672G, T789C, A1017G, C1028G, A1293G, A1517G,...   \n",
       "\n",
       "                                     aa_muts  \\\n",
       "0                [R51K, I124T, T279A, Q384K]   \n",
       "1                                         []   \n",
       "2  [N61D, T124I, D142E, R156K, E228K, P233S]   \n",
       "3                                    [H154Q]   \n",
       "4                      [T336S, K499R, M527I]   \n",
       "\n",
       "                                branch_id  \n",
       "0  <baltic.node object at 0x7f86187a5278>  \n",
       "1  <baltic.node object at 0x7f86187a52b0>  \n",
       "2  <baltic.node object at 0x7f86187a52e8>  \n",
       "3  <baltic.node object at 0x7f86187a5320>  \n",
       "4  <baltic.node object at 0x7f86187a5358>  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"this will generate a multi-index dataframe from the migrations dictionary\"\"\"\n",
    "\n",
    "mutations_df = pd.DataFrame.from_dict({(i,j): mutations_dict[i][j] \n",
    "                           for i in mutations_dict.keys() \n",
    "                           for j in mutations_dict[i].keys()},\n",
    "                       orient='index')\n",
    " \n",
    "mutations_df.reset_index(inplace=True)\n",
    "mutations_df.rename(columns={'level_0': 'tree_number', 'level_1': 'branch_number'}, inplace=True)\n",
    "mutations_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tree_number</th>\n",
       "      <th>branch_number</th>\n",
       "      <th>type</th>\n",
       "      <th>host_switch</th>\n",
       "      <th>host</th>\n",
       "      <th>role</th>\n",
       "      <th>mutations</th>\n",
       "      <th>aa_muts</th>\n",
       "      <th>branch_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>within-wild</td>\n",
       "      <td>no</td>\n",
       "      <td>wild</td>\n",
       "      <td>on</td>\n",
       "      <td>[C72T, G173A, T392C, A856G, C1171A, C1341T]</td>\n",
       "      <td>R51K</td>\n",
       "      <td>&lt;baltic.node object at 0x7f86187a5278&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>within-wild</td>\n",
       "      <td>no</td>\n",
       "      <td>wild</td>\n",
       "      <td>on</td>\n",
       "      <td>[C72T, G173A, T392C, A856G, C1171A, C1341T]</td>\n",
       "      <td>I124T</td>\n",
       "      <td>&lt;baltic.node object at 0x7f86187a5278&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>within-wild</td>\n",
       "      <td>no</td>\n",
       "      <td>wild</td>\n",
       "      <td>on</td>\n",
       "      <td>[C72T, G173A, T392C, A856G, C1171A, C1341T]</td>\n",
       "      <td>T279A</td>\n",
       "      <td>&lt;baltic.node object at 0x7f86187a5278&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>within-wild</td>\n",
       "      <td>no</td>\n",
       "      <td>wild</td>\n",
       "      <td>on</td>\n",
       "      <td>[C72T, G173A, T392C, A856G, C1171A, C1341T]</td>\n",
       "      <td>Q384K</td>\n",
       "      <td>&lt;baltic.node object at 0x7f86187a5278&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>within-wild</td>\n",
       "      <td>no</td>\n",
       "      <td>wild</td>\n",
       "      <td>on</td>\n",
       "      <td>[A9C, C957T]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;baltic.node object at 0x7f86187a52b0&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   tree_number  branch_number         type host_switch  host role  \\\n",
       "0            1              1  within-wild          no  wild   on   \n",
       "0            1              1  within-wild          no  wild   on   \n",
       "0            1              1  within-wild          no  wild   on   \n",
       "0            1              1  within-wild          no  wild   on   \n",
       "1            1              2  within-wild          no  wild   on   \n",
       "\n",
       "                                     mutations aa_muts  \\\n",
       "0  [C72T, G173A, T392C, A856G, C1171A, C1341T]    R51K   \n",
       "0  [C72T, G173A, T392C, A856G, C1171A, C1341T]   I124T   \n",
       "0  [C72T, G173A, T392C, A856G, C1171A, C1341T]   T279A   \n",
       "0  [C72T, G173A, T392C, A856G, C1171A, C1341T]   Q384K   \n",
       "1                                 [A9C, C957T]     NaN   \n",
       "\n",
       "                                branch_id  \n",
       "0  <baltic.node object at 0x7f86187a5278>  \n",
       "0  <baltic.node object at 0x7f86187a5278>  \n",
       "0  <baltic.node object at 0x7f86187a5278>  \n",
       "0  <baltic.node object at 0x7f86187a5278>  \n",
       "1  <baltic.node object at 0x7f86187a52b0>  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "long_form_mutations_df = mutations_df.explode('aa_muts')\n",
    "long_form_mutations_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "# long_form_mutations_df = mutations_df.explode('mutations')\n",
    "# long_form_mutations_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write out to tsv\n",
    "long_form_mutations_df.to_csv(\"/Users/lmoncla/src/h5n1-gwas/data/aa-mutations-on-branches-\"+current_date+\".tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "LHM-basics (python3)",
   "language": "python",
   "name": "lhm-basics"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
